### PR TITLE
DE3198 - Profile date selector

### PR DESCRIPTION
**Pull request by tallented**
* Date selector can be used with both template-based and reactive forms
* Date selector can be provided a default value through the reactive form value
* The parent form value will only be updated if all 3 date components have a valid value
* Profile birth date required validation still works correctly

image

### CODE REVIEW

File: src/app/date-selector/date-selector.component.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+      this.yearsSelection.push(moment().subtract(i, 'y').year());
+      i++;
+    }
+  }
+
+  //
+  // ControlValueAccessor implementation allowing the date selector to interact with parent form
+  //
+
+  registerOnChange(fn: any): void {
```
Comment (Ganonside) [inline src/app/date-selector/date-selector.component.ts]: Is `fn` truly going to be anything or should it just be a function? If it is supposed to be a function it should be typed as a function.

File: src/app/date-selector/date-selector.component.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    this.onChange = fn;
+  }
+
+  registerOnTouched(fn: any): void {
```
Comment (Ganonside) [inline src/app/date-selector/date-selector.component.ts]: see above

File: src/app/date-selector/date-selector.component.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  //
+  // Set up the Month, Day and Year select options
+  //
+
+  populateDateFields() {
+    this.numberOfMonthsSelection = Array.apply(null, {length: 12}).map(function (e, i) { return i + 1; }, Number);
+    this.numberOfDaysSelection = Array.apply(null, {length: 31}).map(function (e, i) { return i + 1; }, Number);
+    this.yearsSelection = [];
+    let i = DateSelectorComponent.MINIMUM_YEARS;
+    while (i <= DateSelectorComponent.NUMBER_YEARS) {
```
Comment (silbermm) [inline src/app/date-selector/date-selector.component.ts]: this loop might be better written using lodash `_.range` and then mapping over the data

```typescript
const yearRange = _.range(DateSelectorComponent.NUMBER_YEARS)
this.yearsSelection = yearRange.map( y => moment().subtract(y, 'y').year())
```

File: src/app/date-selector/date-selector.component.ts [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+  onChangeYear(year: number): void {
+    this.year = year;
+    this.updateExternalDate();
+  }
+
+  // If all of the date components have a valid value, call the registered onChange callback to notify
+  // the parent form of the change in date value
+  updateExternalDate() {
+    if (this.year && this.month && this.day) {
```
Comment (silbermm) [inline src/app/date-selector/date-selector.component.ts]: we should probabaly validate that the actual date is valid.... i.e. not 2/31/2017

**Pull request merged**